sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -4201,6 +4219,73 @@ object SessionWindowing extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Resolves the window_time expression which extracts the correct window time from the
+ * window column generated as the output of the window aggregating operators. The
+ * window column is of type struct { start: TimestampType, end: TimestampType }.
+ * The correct window time for further aggregations is window.end - 1.
+ * */
+object ResolveWindowTime extends Rule[LogicalPlan] {
```

Comment:
can we put it in a new file?

Reply:
Maybe good to move out rules for time window and session window as well while we are here.

ps. It might be even better to consider moving out other rules as well (need help from experts on SQL area) so that no one would think it's proper to add a new rule in this file.

Reply:
Moved window, session_window and window_time resolution to ResolveTimeWindows.scala